File: Problems/Leetcode/3001-4000/3101-3200/3164.py
from collections import Counter
from typing import List


class Solution:
    def numberOfPairs(self, nums1: List[int], nums2: List[int], k: int) -> int:
        d = Counter(x // k for x in nums1 if x % k == 0)
        if not d:
            return 0

        ans = 0
        mx = max(d)
        for x, c in Counter(nums2).items():
            s = sum(d[y] for y in range(x, mx + 1, x))
            ans += s * c
        return ans

File: Problems/Leetcode/3001-4000/3101-3200/3164.cpp
#include <bits/stdc++.h>

using namespace std;

static constexpr int MOD = 1e9 + 7;

class Solution {
public:
    long long numberOfPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        unordered_map<int, int> cnt;
        for (int& v : nums1) {
            if (v % k == 0) {
                cnt[v / k]++;
            }
        }
        if (cnt.empty()) {
            return 0;
        }

        unordered_map<int, int> ump;
        for (int& v : nums2) {
            ump[v]++;
        }

        long long ans = 0;
        int u = ranges::max_element(cnt)->first;
        for (auto& [i, c] : ump) {
            long long s = 0;
            for (int j = i; j <= u; j += i) {
                s += cnt.contains(j) ? cnt[j] : 0;
            }
            ans += s * c;
        }
        return ans;
    }
};

int main() {
    Solution sol;

    return 0;
}